app/src/main/java/com/guru/bluetooth/MainActivity.kt

```diff
@@ -1,194 +1,248 @@
+import android.graphics.Color
 import android.os.Bundle
+import android.widget.AdapterView
+import android.widget.ArrayAdapter
+import android.widget.TextView
 import android.widget.Toast
 import androidx.activity.result.contract.ActivityResultContracts
 import androidx.appcompat.app.AppCompatActivity
-import androidx.core.app.ActivityCompat
-import androidx.core.content.ContextCompat
 import com.guru.bluetooth.databinding.ActivityMainBinding
-import kotlinx.coroutines.CoroutineScope
-import kotlinx.coroutines.Dispatchers
-import kotlinx.coroutines.launch
-import kotlinx.coroutines.withContext
+import com.guru.bluetooth.extensions.showToast
+import com.guru.bluetooth.helper.PermissionsHelper
+import com.guru.bluetooth.server.BluetoothServerController
 
 class MainActivity : AppCompatActivity() {
+
     private lateinit var binding: ActivityMainBinding
-    private val bluetoothAdapter: BluetoothAdapter? by lazy { BluetoothAdapter.getDefaultAdapter() }
-    private var bluetoothDevice: BluetoothDevice? = null
-    private val bluetoothFileTransfer by lazy { BluetoothFileTransfer(this) }
+    private var bluetoothAdapter: BluetoothAdapter? = null
+    private val permissionsHelper = PermissionsHelper(this)
+    private lateinit var pairedDevices: Set<BluetoothDevice>
+    private lateinit var discoveredDevices: ArrayList<BluetoothDevice>
+    private var isInScanningMode: Boolean = false
     private val requestEnableBtLauncher = registerForActivityResult(
         ActivityResultContracts.StartActivityForResult()
     ) { result ->
         if (result.resultCode != RESULT_OK) {
-            Toast.makeText(this, "Bluetooth must be enabled", Toast.LENGTH_SHORT).show()
-            finish()
-        }
-    }
-
-    private val requestDiscoverableLauncher = registerForActivityResult(
-        ActivityResultContracts.StartActivityForResult()
-    ) { result ->
-        if (result.resultCode != RESULT_OK) {
-            Toast.makeText(this, "Device must be discoverable", Toast.LENGTH_SHORT).show()
+            showToast("Bluetooth must be enabled")
             finish()
-        } else {
-            startServer()
         }
     }
 
-    private val requestPermissionLauncher =
-        registerForActivityResult(ActivityResultContracts.RequestPermission()) { isGranted ->
-            if (isGranted) {
-                scanForDevices(this)
-            } else {
-                Toast.makeText(this, "Permission denied", Toast.LENGTH_SHORT).show()
-                finish()
-            }
-        }
-
-    private val deviceClickListener = { device: BluetoothDevice ->
-        bluetoothDevice = device
-        connectToDevice(device)
-    }
-
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         binding = ActivityMainBinding.inflate(layoutInflater)
         setContentView(binding.root)
 
-        setupViews()
+        bluetoothAdapter = BluetoothAdapter.getDefaultAdapter()
```

Comment:
BluetoothAdapter.getDefaultAdapter() is deprecated
Instead, use bluetoothManager.getAdapter()
https://stackoverflow.com/questions/69122978/what-do-i-use-now-that-bluetoothadapter-getdefaultadapter-is-deprecated

Reply:
val bluetoothAdapter =
            (context.getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager).adapter